src/transactions/TransactionFrame.cpp

```diff
@@ -740,6 +740,19 @@ TransactionFrame::applyOperations(SignatureChecker& signatureChecker,
                                   TransactionMeta& outerMeta)
 {
     ZoneScoped;
+    auto& internalErrorCounter = app.getMetrics().NewCounter(
+        {"ledger", "transaction", "internal-error"});
+
+    // We do not want to treat exception as internal-error for non-live ledger
+    // versions. The live ledger version is in the head of the ledger, and
+    // normally should be same as the app.getConfig().LEDGER_PROTOCOL_VERSION,
+    // which should also be same as maxAssertLedgerVersion here. However, right
+    // before a protocol upgrade, the LEDGER_PROTOCOL_VERSION (the software
+    // version) could lead the protocol version by 1, thus we need to offset it
+    // from the maxAssertLedgerVersion.
+    uint32_t maxAssertLedgerVersion =
+        app.getConfig().LEDGER_PROTOCOL_MIN_VERSION_INTERNAL_ERROR_REPORT;
+    std::optional<uint32_t> ledgerVersion;
```

Comment:
I'm not very confident about the use of `std::optional` here. Let's suppose that we don't reach line 765 where it's set (for example, imagine we throw on line 761). We jump down to line 842 in the first catch block, where `ledgerVersion.value()` throws `std::bad_optional_access` (see https://en.cppreference.com/w/cpp/utility/optional/value). This crashes the program.

So there are a few choices:
1. We could check in the `catch` blocks if `ledgerVersion` contains a value. If it doesn't, we just always alert (because it's better to alert when you shouldn't than it is to not alert when you should)
2. We could avoid the use of `std::optional` entirely. To do this, we revert the definition of `uint32_t ledgerVersion` to line 765--like it was before this change--and instead introduce a `bool shouldAlertOnInternalError = true` on this line (`true` for the same reason about always better to alert if you're not sure). Then on line 766 we can set the correct value for `shouldAlertOnInternalError`. This would probably eliminate the need for `maxAssertLedgerVersion` entirely, because it will only be used on that one line.

Second option seems significantly cleaner to me, but I'm open to other ideas.

Reply:
The second option is a great idea.

Reply:
The reason I thought `std::optional` is fine here is because `TransactionFrame::applyOperations` is only called by `apply` under a try-catch block, so if it wasn't set before it will be caught there. But re-thinking that it is probably not a good assumption. 
I agree approach 2 is significantly better and cleaner. Will make the change.

Reply:
The problem with approach 2 is we'll still need the ledger version in the two `catch` blocks. For that reason, I'm going with option 1. Please see changes.

Reply:
So you need the ledgerVersion in the catch block purely for logging in approach 2. Do we actually need this? I don't think we care about which old protocol version it is, just that it's less than `LEDGER_PROTOCOL_MIN_VERSION_INTERNAL_ERROR_REPORT`.

Reply:
Yes it is for logging only. It was mentioned in the issue so I interpreted it as a requirement.

Reply:
The issue mentions logging the protocol version as a suggestion. I think the initial idea from Nico in that issue is sufficient ```"add information in the logs that this is for an older protocol version, maybe reduce to INFO as opposed to ERROR"```. What do you think?

Reply:
I agree, in that case, probably best to leave the version number out of the logging in favor of cleaner code. (Approach 2)

Reply:
See fix.